File: product/server/backoffice/src/main/java/io/codekvast/backoffice/service/MailSender.java
/*
 * Copyright (c) 2015-2019 Hallin Information Technology AB
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */
package io.codekvast.backoffice.service;

import lombok.Getter;
import lombok.RequiredArgsConstructor;

/**
 * @author olle.hallin@crisp.se
 */
public interface MailSender {

    @RequiredArgsConstructor
    @Getter
    enum Template {
        WELCOME_COLLECTION_HAS_STARTED("Welcome to Codekvast!");

        private final String subject;
    }

    /**
     * Sends an email using a named template to a given email address.
     *
     * @param template     The template to use.
     * @param customerId   The customer ID.
     * @param emailAddress The email address to send to.
     */
    void sendMail(Template template, Long customerId, String emailAddress);
}

File: product/server/backoffice/src/main/java/io/codekvast/backoffice/service/MailSender.kt
/*
 * Copyright (c) 2015-2022 Hallin Information Technology AB
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */
package io.codekvast.backoffice.service

/** @author olle.hallin@crisp.se
 */
interface MailSender {
    /**
     * Sends an email using a named template to a given email address.
     *
     * @param template The template to use.
     * @param emailAddress The email address to send to.
     * @param args The arguments to the template.
     */
    fun sendMail(template: Template, emailAddress: String, vararg args: Any)

    enum class Template(val subject: String) {
        WELCOME_TO_CODEKVAST("Welcome to Codekvast!");
    }
}
